kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/ResolverAAImpl.kt

```diff
@@ -865,10 +865,25 @@ class ResolverAAImpl(
                     )
                 }
                 analyze {
-                    // TODO: recursive substitution does not solve fix point problem for signatures, probably needs fix in upstream.
                     buildSubstitutor {
                         fillInDeepSubstitutor(containing.type, this@buildSubstitutor)
-                    }.let { (function as KSFunctionDeclarationImpl).ktFunctionSymbol.substitute(it) }.let {
+                    }.let {
+                        // recursively substitute to ensure transitive substitution works.
+                        // should fix in upstream as well.
+                        // for functions we need to test both returnType and value parameters converges.
+                        var funcToSub = (function as KSFunctionDeclarationImpl).ktFunctionSymbol.substitute(it)
+                        var next = funcToSub.substitute(it)
+                        while (
```

Comment:
The complexity of this logic and also its dependency on the substitutor makes me worried whether it always converges. Can you add a counter to throw in case it doesn't?

Reply:
Sounds fair to me, I thought about the converge issue before but can't think about a fair number to exit.

Reply:
tentatively put 100 as a safe value.